037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,177 @@
+   * If auto-approval and criteria are not met, the `KafkaRebalance` custom resource status will reflect, by using a warning condition, that the proposal was not approved automatically and the reason for that.
+5. Only when the proposal is approved and rebalancing is complete, the cluster operator can continue to scale down the cluster by deleting the brokers.
+
+In this use case, the actual scaling down (deleting the brokers) is blocked on the proposal to be approved and the rebalancing done.
+To fix this kind of issue, the user should change the rebalancing configuration and allow the cluster operator to reconcile again for getting a new optimization proposal, running a rebalance and finally unlocking the scaling down.
+
+## Affected/not affected projects
+
+This proposal impacts the Strimzi Cluster Operator only and mostly the `KafkaRebalanceAssemblyOperator`.
+
+## Compatibility
+
+The auto-rebalancing on scaling up and down is optional.
+The user has to specify the `spec.kafka.autoRebalance` on the `Kafka` custom resource for the desire of having the operator doing the rebalancing automatically.
+Without that addition, the rebalancing can be done with the manual two step process as today.
+
+## Rejected alternatives
+
+When the `spec.kafka.autoRebalance` is specified, the cluster operator can create an “empty” `KafkaRebalance` custom resource automatically.
+By “empty” custom resource we mean that the default Cruise Control configuration will be used for goals and all other options.
+The only fields that the cluster operator has to specify, when creating the `KafkaRebalance` custom resource, based on scaling up or down, are the:
+
+* `spec.mode` : should be `add-brokers` or `remove-brokers` if scaling up or down.
+* `spec.brokers` : with the list of brokers to add/remove.
+
+The drawback of this solution is that if the user wants to customize the rebalancing, he has to use the `spec.cruiseControl.config` field in the Kafka custom resource. It’s important to mention that it affects the overall Cruise Control configuration so that any other “empty” rebalancing request, not related to scaling, will use the same options.
+
+To allow customization, a potential solution would be to add rebalancing related configuration (goals, and so on) under the `spec.kafka.autoRebalance` directly.
+It would be a kind of copy of the fields from the `KafkaRebalance` custom resource excluding the mode and brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      mode: [running, scaling] # or just one of the two
+      goals:
+        - CpuCapacityGoal
+        - NetworkInboundCapacityGoal
+        - DiskCapacityGoal
+        - RackAwareGoal
+        - MinTopicLeadersPerBrokerGoal
+        - NetworkOutboundCapacityGoal
+        - ReplicaCapacityGoal
+      skipHardGoalCheck: true
+      # ... other rebalancing related configuration
```

Comment:
What else would be here? I think it is good to clarify this into more detail to understand how big it would possibly be. It is important part to decide if `KAfkaRebalanceTemplate` is better than this or not (i.e. to balance CRD size with usability and readability).

Reply:
I added a link to the Strimzi documentation with the current options we allow to configure today but with possibility they will increase in the future.